[MVC - Step 1] 슬로스 미션 제출합니다.

soojinroh opened the pull request:
안녕하세요 현구님 슬로스입니다!

이번 미션에서는 레거시 코드를 최대한 수정하지 않으면서 레거시 MVC와 애노테이션 기반 MVC를 통합하는 데에 중점을 두고 미션을 진행했습니다.

어차피 나중에는 애노테이션 기반의 MVC를 활용하는 것으로 모두 리팩토링할 것이라 생각하면서 미션을 진행하다보니 자꾸 레거시 코드를 배려하지 않는(?) 방향으로 통합하게 되는 것 같아 레거시 코드의 양이 방대할 수 있다는 것을 염두에 두고 미션을 진행하는 점이 어려웠던 것 같습니다.

잘 부탁드립니다!

soojinroh committed:
feat : Class reflection 구현

soojinroh committed:
feat : Junit3TestRunner 구현

soojinroh committed:
feat : Junit4TestRunner 구현

soojinroh committed:
feat : 인스턴스의 Field 접근 Reflection 구현

soojinroh committed:
feat : ReflectionTest 구현

인자가 있는 생성자를 reflection으로 생성

soojinroh committed:
feat : ReflectionsTest showAnnotationClass 테스트 구현

soojinroh committed:
feat : AnnotationHandlerMapping 클래스 구현

soojinroh committed:
feat : RequestMappingHandlerMapping 클래스 구현

HTTP 요청에 따라 LegacyHandlerMapping클래스와 AnnotationHandlerMapping클래스 중 요청에 해당하는 Handler를 찾아 실행시키고 ViewName을 리턴함

soojinroh committed:
feat : RequestMappingHandlerMapping 적용

기존에 DispatcherServlet에서 MannualHandlerMapping클래스를 주입받아 사용하던 것을 Legacy와 리팩토링한 Handler를 모두 동작시킬 수 있게끔 RequestMappingHandlerMapping을 받아 사용하는 것으로 수정

soojinroh committed:
refactor : 클래스명 변경

HandlerMapping 클래스를 LegacyHandlerMapping으로 이름 변경

soojinroh committed:
refactor : Legacy Controller에 RequestMapping Annotation 적용

soojinroh committed:
style : 주석 삭제

soojinroh committed:
refactor : HandlerKey 클래스 생성자 수정

soojinroh committed:
refactor : RequestMappingHandlerMapping의 handle method return 타입 변경

기존에 String으로 ViewName을 return하던 것에서 ModelAndView를 return하는 것으로 수정.
View 인터페이스를 구현한 JspView클래스를 만들어 View 정보를 담아 ModelAndView에 담아 Return함.

soojinroh committed:
test : CreateUserController test 추가

soojinroh committed:
test : test 코드 컨벤션 수정

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/RequestMappingHandlerMapping.java:
```diff
@@ -0,0 +1,34 @@
+    private AnnotationHandlerMapping annotationMapping;
+    private LegacyHandlerMapping legacyMapping;
+
+    public RequestMappingHandlerMapping(LegacyHandlerMapping legacyMapping) {
```
어노테이션 외에 새로운 방식의 HandlerMapping이 추가될때마다 인스턴스 변수가 늘어나고 생성자를 수정해야겠네요
HandlerMapping 인터페이스를 레거시 핸들러 매핑만 쓰지 말고 어노테이션 핸들러 매핑에서도 구현(implements) 하고, dispatcher servlet에서 리스트로 전달 받아서 for each 돌리면서 initialize() 하는 구조로 바꾸면 어떨까요?

soojinroh replied:
피드백 주신 내용 참고해서 리팩토링 하다보니 Adapter pattern을 적용하게 되었고 view 부분도 건들여서 시간이 좀 오래 걸렸습니다ㅜ

ReqeustMappingHanderMapping이 AnnotationHandlerMapping과 ManualHandlerMapping(LegacyHandlerMapping)을 가지고 있는 건 피드백 주신 내용과 같은 한계가 있는 것 같아 ManualHandlerAdapter와 AnnotationHandlerMapping이 HandlerResolver를 구현하는 것으로 수정해보았습니다!

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -16,10 +24,30 @@ public AnnotationHandlerMapping(Object... basePackage) {
 
+    private void mappingMethod(Reflections reflections) {
+        reflections.getMethodsAnnotatedWith(RequestMapping.class).stream()
+                .filter(method -> Arrays.stream(method.getParameterTypes())
```
1. 이중 for문이 되면서 코드 읽기가 어려워지면 유지보수나 리팩토링하기 힘들 수 있어요!
2. 컨트롤러에 있는 메서드들이 request, response만 처리하도록 구현할 필요는 없어보여요. 스프링 생각해보면 RequestParam, RequestBody, PathVariable 같은 다양한 방식을 제공하고 있으니 메서드에 RequestMapping를 사용하고 있는지만 체크해도 충분해보여요

soojinroh replied:
피드백 주신 두가지 포인트 모두 참고해서 수정하였습니다!

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -16,10 +24,30 @@ public AnnotationHandlerMapping(Object... basePackage) {
+                .forEach(method -> {
+                    RequestMapping mapping = method.getDeclaredAnnotation(RequestMapping.class);
+                    HandlerExecution handlerExecution = new HandlerExecution();
+                    handlerExecution.setMethod(method);
```
setter 사용 금지!! setter 쓰면 캡슐화가 깨져요
HandlerExecution 생성자를 추가하고 method, newInstance로 생성한 객체를 전달해보면 어떨까요

soojinroh replied:
피드백 반영했습니다!

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -16,10 +24,30 @@ public AnnotationHandlerMapping(Object... basePackage) {
+                    } catch (InstantiationException | IllegalAccessException e) {
+                        logger.debug(e.getMessage());
+                    }
+                    handlerExecutions.put(HandlerKey.of(mapping.value(), mapping.method()), handlerExecution);
```
HandlerKey를 팩토리 메서드로 생성할 필요는 없어보이네요

soojinroh replied:
피드백 반영했습니다!

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/JsonView.java:
```diff
@@ -9,4 +9,9 @@
+
+    @Override
+    public String getViewName() {
+        return null;
```
json view에서는 의미없는 메서드 같은데 인터페이스로 추상화할 필요가 있었을까요?

soojinroh replied:
view 부분은 2단계 미션이라고 생각해서 좀 더 깊게 생각 못했던 것 같습니다.

피드백 주신 것처럼 json에서는 viewName getter가 필요 없다고 생각되어 View interface에서 해당 메서드를 삭제하였습니다!

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/JspView.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    @Override
+    public void render(Map<String, ?> model, HttpServletRequest request, HttpServletResponse response) throws Exception {
```
렌더링 관련 로직이 빠진것 같은데... 의도한걸까요?

soojinroh replied:
수정했습니다!

kang-hyungu reviewed (changes requested):
안녕하세요 슬로스님 리뷰어 강현구입니다
mvc 프레임워크 구현 미션 진행하느라 수고하셨어요
레거시 코드도 유지하면서 어떻게 리팩토링하면 좋을지 피드백 남겼으니 참고하셔서 구현해보세요
구조 바꾸면서 더 궁금한 점이 생길텐데 빠르게 피드백 요청주시거나 dm 주시면 답변드릴게요!

soojinroh committed:
refactor : HandlerExcution setter 삭제 후 생성자로 상태 전달

soojinroh committed:
refactor : HandlerKey 팩토리 메서드 제거

soojinroh committed:
refactor : AnnotationHandlerMapping의 mappingHandler 메서드 수정

1. 이중 for문 제거
2. 메서드의 파라미터가 아닌 애노테이션을 확인해서 매핑하는 방식으로 변경

soojinroh committed:
refactor : Controller 구현 Handler를 RequestMapping 애노테이션 설정 방식으로 변경

soojinroh committed:
test : ReqeustMapping Annotation방식으로 변경된 클래스 테스트 코드 추가

soojinroh committed:
test : Jackson Library 학습

Jackson Library를 이용해 Java Object를 Json으로, Json String을 Java Object로 변환하는 방법 학습

kang-hyungu commented:
https://github.com/woowacourse/jwp-mvc/pull/22#discussion_r331735364
관련해서 추가 코멘트 남겨요!

레거시 코드에 새로운 기술(어노테이션 기반 컨트롤러)을 추가하다보면 지금처럼 if/else가 계속해서 생길텐데요
if/else가 늘어날 것으로 예상되는 부분에 자바의 다형성을 활용해서 interface를 추가하고 확장 가능한 구조로 개선해보아요!

```java
public interface HandlerAdapter {
    boolean supports(Object handler);

    ModelAndView handle(HttpServletRequest request, HttpServletResponse response, Object handler) throws Exception;
}
```

soojinroh committed:
refactor : LegacyHandlerMapping 클래스명 변경

LengacyHandlerMapping에서 HandlerMapping으로 클래스명 변경

soojinroh committed:
feat : JsonView render 기능 구현

Jackson 라이브러리를 이용하여 JAVA Object를 Json 형식의 String으로 변환

soojinroh committed:
refactor : HandlerMapping에 Adapter 패턴 적용

AnnotationHandlerMapping과 HandlerMappingAdapter가 Adapter를 구현(implements)하고 HandlerMappingAdapter가 ManualHandlerMapping을 가지게 함으로써 adapter패턴을 적용함.

soojinroh committed:
feat : ControllerScanner, RequestMappingScanner 구현

soojinroh committed:
feat : Adapter 구현에 따른 수정

soojinroh committed:
refactor : Legacy Controller를 Annotation 방식으로 리팩토링

soojinroh committed:
refactor : package 이동

soojinroh committed:
feat : ViewResolver 구현

soojinroh committed:
refactor : 생성자 및 getter, setter 추가/수정

soojinroh committed:
refactor : 컨벤션 수정, 클래스 위치 및 이름 변경에 따른 import문 변경

soojinroh committed:
feat : catch문에서 RuntimeExeption throw하는 것으로 수정

soojinroh committed:
feat : catch문에서 RuntimeException throw하는 것으로 수정

soojinroh committed:
refactor : 사용하지 않는 getter 삭제

soojinroh committed:
refactor : RequestMapping의 scan 메서드 수정

indent문제 해결을 위해 method 추출

soojinroh committed:
refactor : 변수명 변경

HandlerAdapter 인터페이스를 HandlerResolver로 변경함에 따라 List<HandlerResolver>의 변수명을 변경함

soojinroh committed:
refactor : 변수명 변경 및 메서드 추출

로직상 변경 없이 변수명 변경과 메서드 추출 리팩토링

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/DispatcherServlet.java:
```diff
@@ -1,57 +1,77 @@
+                ;
+    }
+
+    private ModelAndView handle(HttpServletRequest req, HttpServletResponse resp, HandlerExecution handlerExecution) {
```
```suggestion
    private ModelAndView handle(HttpServletRequest req, HttpServletResponse resp, HandlerExecution handlerExecution) {
        try {
            return handlerExecution.handle(req, resp);
        } catch (Exception e) {
            logger.debug("message : {}, cause : {}", e.getMessage(), e.getCause());
            throw new RuntimeException(e);
        }
    }
```
이렇게 하면 로컬 변수를 선언하지 않아도 될 것 같네요

soojinroh replied:
감사합니다 피드백 적용했습니다!

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/handler/RequestMappingHandlerExecution.java:
```diff
@@ -0,0 +1,23 @@
+
+    public ModelAndView handle(HttpServletRequest request, HttpServletResponse response) throws Exception {
+        ModelAndView result = (ModelAndView) method.invoke(target, request, response);
+        return new ModelAndView(result.getViewName(), result.getModel());
```
ModelAndView를 다시 감쌀 필요가 있을까요?

soojinroh replied:
지금 보니 왜 저렇게 했는지 제가 봐도 이상하네요...바로 리턴하는 것으로 수정하였습니다!

soojinroh replied:
지금 보니 왜 저렇게 했는지 제가 봐도 이상하네요...바로 리턴하는 것으로 수정하였습니다!

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/handlerresolver/AnnotationHandlerMapping.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.Map;
+import java.util.Set;
+
+public class AnnotationHandlerMapping implements HandlerResolver {
```
HandlerResolver에 handler mapping 역할과 handler adapter 역할이 섞였네요
두 개를 분리하면 HandlerExecution를 컨트롤러 인터페이스용과 어노테이션용으로 각각 만들 필요가 없어질거에요

soojinroh replied:
위에서 지적해주신 것처럼 HandlerMapping도 레거시라고 생각해서 getHandler의 return type을 수정하지 않았었습니다. 그러다보니 지금과 같이 구조가 복잡해지고 객체의 역할이 분명하게 나뉘지 못한 것 같습니다.

그래서 피드백 주신 것처럼 getHandler의 return type을 Object로 수정한 뒤, HandlerResolver를 없애고 HandlerResolver가 가지고 있던 기능을 HandlerMapping과 HandlerAdapter로 나누어 구현해보았습니다!

soojinroh replied:
위에서 지적해주신 것처럼 HandlerMapping도 레거시라고 생각해서 getHandler의 return type을 수정하지 않았었습니다. 그러다보니 지금과 같이 구조가 복잡해지고 객체의 역할이 분명하게 나뉘지 못한 것 같습니다.

그래서 피드백 주신 것처럼 getHandler의 return type을 Object로 수정한 뒤, HandlerResolver를 없애고 HandlerResolver가 가지고 있던 기능을 HandlerMapping과 HandlerAdapter로 나누어 구현해보았습니다!

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/handler/HandlerExecution.java:
```diff
@@ -1,10 +1,11 @@
-    public ModelAndView handle(HttpServletRequest request, HttpServletResponse response) throws Exception {
-        return null;
-    }
+public interface HandlerExecution {
```
인터페이스로 바꿀 필요는 없어 보이네요

soojinroh replied:
클래스로 다시 변경하였습니다

soojinroh replied:
클래스로 다시 수정하였습니다!

soojinroh replied:
클래스로 다시 수정하였습니다!

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/handlerresolver/HandlerResolver.java:
```diff
@@ -0,0 +1,14 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public interface HandlerResolver {
```
핸들러 매핑과 핸들러 어댑터 역할을 동시에 가지면서 구조가 복잡해진 것 같네요
두 개로 나누면 코드 복잡도를 더 낮출 수 있을거에요

soojinroh replied:
HandlerMapping과 HandlerAdapter를 나누는 것으로 수정하였습니다!

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/view/JsonView.java:
```diff
@@ -0,0 +1,24 @@
+        ObjectMapper objectMapper = new ObjectMapper();
+
+        for (String key : model.keySet()) {
+            writer.write(objectMapper.writeValueAsString(model.get(key)));
```
아래 조건을 만족하도록 수정해보세요!
```
Map에 담긴 model 데이터가 1개인 경우 value 값을 반환, 2개 이상인 경우 Map 자체를 JSON으로 변환해 반환한다.
```

soojinroh replied:
수정했습니다!

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/view/ModelAndView.java:
```diff
@@ -1,20 +1,29 @@
 
 public class ModelAndView {
-    private View view;
+    private Object view;
```
View로 다시 바꾸는게 좋겠네요

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/view/ModelAndView.java:
```diff
@@ -1,20 +1,29 @@
     public ModelAndView() {
     }
 
+    public ModelAndView(String viewName) {
```
이 생성자도 없애는게 좋겠어요
컨트롤러에서 JspView 같은 객체로 만들어서 전달해주면 되지 않을까요?

soojinroh replied:
JspView, RedirectView, JsonView를 컨트롤러에서 만들어서 전달해주는 구조로 수정하였습니다! 그러다보니 ViewResolver의 역할이 사라진 것 같아 ViewResolver를 삭제하였습니다.

kang-hyungu commented on nextstep-mvc/src/test/java/nextstep/mvc/tobe/JsonViewTest.java:
```diff
@@ -59,4 +62,18 @@ void render_over_two_element() throws Exception {
+
+        view.render(model, request, response);
+
+        assertThat(response.getContentType()).isEqualTo(MediaType.APPLICATION_JSON_UTF8_VALUE);
```
json 결과가 의도한대로 나왔는지 테스트 추가해보세요!

soojinroh replied:
테스트 추가했습니다!

kang-hyungu commented:
HandlerMapping에서 아래 코드로 수정해보면 HandlerExecution을 각각 구현할 필요 없이 클래스 하나로 처리할 수 있어요
아마 이부분도 레거시 코드에 포함된다고 생각하셔서 HandlerResolver를 추가하신 것 같네요

```java
Object getHandler(final HttpServletRequest request);
```

kang-hyungu reviewed (approved):
코드 구현하느라 수고하셨어요!
코드 구조를 좀더 바꾸면 좋을 것 같은 부분에 코멘트 남겼어요
확인해보시고 다음 미션 진행하면서 적용해보세요!
